src/main/java/at/laborg/briss/gui/DrawableCropRect.java

```diff
@@ -50,37 +63,86 @@ public DrawableCropRect(int x, int y, int width, int height) {
-        cornerLowerRight.setLocation(x + width - halfCornerDimension, y + height - halfCornerDimension);
+        cornerUpperLeft.setLocation(x, y);
+        cornerUpperRight.setLocation(x + width, y);
+        cornerLowerLeft.setLocation(x, y + height);
+        cornerLowerRight.setLocation(x + width, y + height);
     }
 
-    public void draw(Graphics2D g2) {
+    public boolean hasEnoughSpaceForHandles() {
+        return !(getWidth() < 2 * DrawableCropRect.CORNER_DIMENSION)
+            && !(getHeight() < 2 * DrawableCropRect.CORNER_DIMENSION);
+    }
+
+    private Font scaleFont(String text, FontMetrics fontMetrics) {
+
+        int size = MergedPanel.BASE_FONT.getSize();
+        float width = fontMetrics.stringWidth(text);
+        float height = fontMetrics.getHeight();
+        if (width == 0 || height == 0)
+            return MergedPanel.BASE_FONT;
+        float scaleFactorWidth = (float) (this.width * 0.95 / width);
+        float scaleFactorHeight = (float) (this.height * 0.95 / height);
+        float scaledWidth = (scaleFactorWidth * size);
+        float scaledHeight = (scaleFactorHeight * size);
+        return MergedPanel.BASE_FONT.deriveFont((scaleFactorHeight > scaleFactorWidth) ? scaledWidth : scaledHeight);
+    }
+
+    public void draw(Graphics2D g2, int cropRectNumber, FontMetrics fontMetrics) {
+
+        g2.setComposite(SMOOTH_NORMAL);
+        if (hasEnoughSpaceForHandles()) {
+            g2.setColor(Color.BLUE);
+        } else {
+            g2.setColor(Color.RED);
+        }
+        g2.fill(this);
+        g2.setColor(Color.BLACK);
+        g2.setFont(scaleFont(String.valueOf(cropRectNumber + 1), fontMetrics));
+        g2.drawString(String.valueOf(cropRectNumber + 1), this.x, this.y + this.height);
 
         if (isSelected()) {
-            g2.setPaintMode();
-            g2.setColor(Color.WHITE);
-            g2.fill(cornerUpperLeft);
-            g2.fill(cornerUpperRight);
-            g2.fill(cornerLowerLeft);
-            g2.fill(cornerLowerRight);
-            g2.setColor(Color.BLACK);
-            g2.draw(cornerUpperLeft);
-            g2.draw(cornerUpperRight);
-            g2.draw(cornerLowerLeft);
-            g2.draw(cornerLowerRight);
+            drawSelectionOverlay(g2, fontMetrics);
+
+            if (hasEnoughSpaceForHandles()) {
+                drawResizeHandles(g2);
+            }
         }
     }
 
+    private void drawResizeHandles(Graphics2D g2) {
+        g2.setPaintMode();
+        cornerUpperLeft.draw(g2);
+        cornerUpperRight.draw(g2);
+        cornerLowerLeft.draw(g2);
+        cornerLowerRight.draw(g2);
+    }
+
+
+    private void drawSelectionOverlay(Graphics2D g2, FontMetrics fontMetrics) {
+        g2.setComposite(XOR_COMPOSITE);
+        g2.setColor(Color.BLACK);
+
+        g2.setStroke(SELECTED_STROKE);
+        g2.draw(this);
+
+        // display crop size in millimeters
+        int w = Math.round(25.4f * this.width / 72f);
```

Comment:
Avoid magic constants here as well.

Reply:
Actually I just moved the these magic constants just from the panel but I tried to come up with meaningful names provided I have understood them correctly 😄